### ISSUE TITLE
Jenkins: Introduction

### ISSUE THREAD

**Issue by fesiib** [temirrr/Priora#1]
Learn how to design and implement plugin for Jenkins

**Comment by temirrr** [comment]
This is our initial architecture: <https://drive.google.com/file/d/15u1eNNXVnifZPrR_Dl-3SSyvQDT_dOg5/view?usp=sharing> . Feel free to modify it.

Also, we can consider points such as:
- How will commits trigger the (re)execution of all pipeline.
- Find out how to measure the test execution time and the test result

**Comment by fesiib** [comment]
Useful Jenkins Plugins:
[parallel-test-executor-plugin](https://github.com/jenkinsci/parallel-test-executor-plugin) - uses the execution time of each test from last time, to split the tests into chunks that will be executed in parallel. Also, I suspect that both commit trigger and execution time calculator are used here.
[qtest-plugin](https://github.com/jenkinsci/qtest-plugin) - collects JUnit test results and submit them as Test Runs and Test Logs to qTest. Plugin's collection and transfer of the JUnit test results can be helpful.

**Comment by fesiib** [comment]
Some important points from [Jenkins/HostingPlugins](https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins#HostingPlugins-HostingPlugins-AddingMaintainerInformation), we have to do these preferably at the beginning:

1) Give the plugin a sensible ID which does not include "jenkins" or "plugin". 
- This is the <artifactId> tag in pom.xml since we're using Maven 

2) Give the plugin a descriptive but short name; 
- This is the <name> tag in pom.xml

3) We are also encouraged to give the plugin a short one-line description;
- This is the <description> tag in pom.xml

2 and 3 will be shown in Update Center

4) Have at least a small README explaining what the plugin does/is about.

5)  Set parent of groupId org.jenkins-ci.plugins. The groupId for the project (outside of the parent declaration) should be set to io.jenkins.plugins

6) Specify open-source license

7) Have a jenkins-ci.org account.
- This gives you access to upload your plugin, document your plugin on this wiki, and track issues in JIRA. It's also required to request hosting.

**Comment by fesiib** [comment]
1) Find test related Jenkins extension points?

2) Hudson?

**Comment by temirrr** [comment]
I found out that Jenkins has `cron` syntax to do repo polling at regular time intervals - but this is not what we want. We rather want **commits** to trigger the pipeline. Additionally, `cron` syntax can only poll every 2 minutes (actually, every 1 minute, but there is a popular Jenkins bug *JENKINS-22129* about that).
https://www.jenkins.io/doc/book/pipeline/syntax/#cron-syntax
https://groups.google.com/forum/#!topic/jenkinsci-users/cAr7OMCct2U

So, I guess the best we can do is manually triggering the pipeline execution ourselves. We can do that by using `post-commit` hook in Git repository. That hook, combined with `Git Plugin`, can trigger the execution of the pipeline.
https://stackoverflow.com/questions/12794568/how-to-configure-git-post-commit-hook

The only problem is:
> and if that finds a change worth a build, a build will be triggered in turn.

So, the `post-commit` hook will trigger the Jenkins. And Jenkins will start by checking the code for meaningful changes, which affect the build.
As we discussed yesterday with @fesiib , it's too hard to add some meaningful changes to the code in the automated way. But I guess we can keep adding some pseudo-meaningful changes such as `printf("Nothing meaningful");`. I hope that counts like smth that needs to be actually compiled/built. So we need to find a way to write those kinds of lines to the program by using `echo ... > ...`, or some other way.

**Comment by temirrr** [comment]
@fesiib @amanzholov8 
Last time, we considered one job sorter plugin for Jenkins. I still don't exactly get what the "Job" is, but I think that we'd better set the test case priority with the JUnit. But I think we can do that later and we all should start coding smth related to Jenking pipeline/plugins. What do you think about that?

**Comment by fesiib** [comment]
So, as a first step, I found out that JUnit plugin in Jenkins is very useful for us. Along supplying testing information in the graphic picture, it has many methods that may help us in our development. To be precise, JUnit plugin already has implemented:
1) Test Info Storage (#success, time consumed)
2) Test Info Retrieval
3) Test Info Update (it is done automatically by Jenkins AFAIK)

Thus, we have to implement:
1) "Test Order Setter".
2) NSGA-II, Random, and Greedy.

However, further analysis needed, so I am working on how to set the test order and trigger testing phase.

**Comment by amanzholov8** [comment]
If you have written any code, could you push what you've done to this repo so that we can also take a look and possibly contribute to what you've already done?

**Comment by fesiib** [comment]
Pushed Environment for Jenkins Plugin development and testing. Following are tasks that we have to complete:

- [ ] Change the sample maven project to benchmark

- [ ] Implement Script to Reorder Tests

- [ ] Implement Plugin that produces desired order using Algorithm

- [ ] Documentation of Plugin

- [ ] Evaluation

- [ ] Report
